### ISSUE TITLE
SSH to a VM via WSL doesn't work

### ISSUE THREAD

**Issue by Zemimmo** [microsoft/WSL#7971]
Hello, I've the following issue: I'm trying to connect via WSL (using ssh) to a VM. 
If I set the VM with bridge network it works properly just using the command "ssh 192.168.1.xxx" (the IP address that the bridge DHCP sets for my VM) 
For my project it should be better to put the VM in NAT Network. I configured NatNetwork with the port forwarding as follows:
Name: SSH
Protocol: TCP
Host's IP: 
Host's Port: 5022
Guest's IP: 10.0.2.xxx (the IP address that the NAT DHCP sets for my VM)
Guest's Port: 22
Once I use these settings the output of the WSL command: "  ssh user@127.0.0.1 -p 5022 -vvv" is:

OpenSSH_8.2p1 Ubuntu-4ubuntu0.3, OpenSSL 1.1.1f  31 Mar 2020
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug2: resolve_canonicalize: hostname 127.0.0.1 is address
debug2: ssh_connect_direct
debug1: Connecting to 127.0.0.1 [127.0.0.1] port 5022.
debug1: connect to address 127.0.0.1 port 5022: Connection refused
ssh: connect to host 127.0.0.1 port 5022: Connection refused

If i use the command via Prompt or PowerShell it works fine. It just doesn't work on WSL.
Can anyone help me?

**Comment by craigloewen-msft** [comment]
Please see this docs page about connecting via the local network  (You may need to use port forwarding):
https://docs.microsoft.com/en-us/windows/wsl/networking#accessing-a-wsl-2-distribution-from-your-local-area-network-lan

And overall this issue seems best tracked here as a duplicate:
https://github.com/microsoft/WSL/issues/4150

Thank you for filing!

**Comment by Zemimmo** [comment]
No one of the link seems to be helpfull. My issue Is not to Accessing a WSL
2 distribution from my local area network and I've not a service in my
Ubuntu on WSL that has to be accessibile from anywhere on my local network.
These issue sounds to me completely different. I've a Virtual Machine
reachable from any other device in my local machine via SSH but not from
wsl. Wsl cannot connect via SSH if I use Port forwarding, It can access
only setting up the Virtual Machine in Bridge mode

Il Mar 25 Gen 2022, 23:36 Craig Loewen ***@***.***> ha
scritto:

**Comment by hemant3434** [comment]
experiencing the same errors @Zemimmo , were you able to find a fix for this?

**Comment by zmweske** [comment]
If this helps, you should use ssh.exe from inside of WSL or try ssh-ing from powershell (on the Windows side). The 127.0.0.1 you are trying to access is the windows localhost. If you use ssh.exe, you are running the windows ssh and it may connect properly. There may be a way to connect to the guest vm directly with WSL ssh by connecting to the WSL network host (found under the WSL adapter of `ipconfig` in windows) instead of 127.0.0.1, but the port forwarding may need to be changed as well. @Zemimmo @hemant3434

**Comment by ericfrederich** [comment]
@craigloewen-msft, would you consider reopening this?

This issue is about connecting from WSL2 into a NAT'd VM service.

It seems #4150 is about connecting from anywhere on the network into a WSL2 service.

**Comment by tungstengmd** [comment]
@craigloewen-msft please reopen this because this problem is about connecting *to* a vm *from* wsl, which isn't possible for god knows what reason

**Comment by manov555** [comment]
@tungstengmd If you are using Windows, you need to specify the IP Address of the Ethernet-Adapter vEthernet (WSL (Hyper-V firewall)) in port forwarding, then it works. Notice: dont use root for login, just normal user

**Comment by tungstengmd** [comment]
https://medium.com/codemonday/ssh-to-virtualbox-host-only-adapter-interface-from-wsl2-machine-f17fc0978edd it seems this helps, just remove all port forwarding rules i suppose

**Comment by kjetillll** [comment]
I also struggled to ssh from WSL2 to a virtual machine on the same computer. What eventually worked for me was to replace the `ssh` command with `ssh.exe -p 2222 vmusername@127.0.0.1` or `/mnt/c/Windows/System32/OpenSSH/ssh.exe`. Before that I added a Port Forwarding rule from host port 2222 to guest port 22 in Virtual Box network settings. You might also have to tell a firewall (ufw perhaps) inside the vm to open 22, if a firewall is active there. From what I undertand the WSL2 is a "lightweight virtual machine" itself which has its own network and its own 127.0.0.1 (localhost), kind of. Run `which ssh` and `which ssh.exe`, or `ssh --version` and `ssh.exe --version` to see the difference if there are any.
